sign/bls/bls.go

```diff
@@ -0,0 +1,54 @@
+package bls
+
+import (
+	"crypto/cipher"
+	"errors"
+
+	"github.com/dedis/kyber"
+	"github.com/dedis/kyber/pairing"
+)
+
+// NewKeyPair creates a new BLS signing key pair. The private key x is a scalar
+// and the public key X is a point on curve G2.
+func NewKeyPair(suite pairing.Suite, random cipher.Stream) (kyber.Scalar, kyber.Point) {
+	x := suite.G2().Scalar().Pick(random)
+	X := suite.G2().Point().Mul(x, nil)
+	return x, X
+}
+
+// Sign creates a BLS signature s = x * H(m) on a message m using the private
+// key x. The signature s is a point on curve G1.
+func Sign(suite pairing.Suite, x kyber.Scalar, msg []byte) []byte {
+	HM := hash(suite, msg)
+	xHM := HM.Mul(x, HM)
+	s, _ := xHM.MarshalBinary()
+	return s
+}
+
+// Verify checks the given BLS signature s on the message m using the public
+// key X by verifying that the equality e(H(m), X) == e(H(m), x*B2) ==
+// e(x*H(m), B2) == e(s, B2) holds where e is the pairing operation and B2 is
+// the base point from curve G2.
+func Verify(suite pairing.Suite, X kyber.Point, msg, sig []byte) error {
+	HM := hash(suite, msg)
+	left := suite.Pair(HM, X)
+	s := suite.G1().Point()
+	if err := s.UnmarshalBinary(sig); err != nil {
+		return err
+	}
+	right := suite.Pair(s, suite.G2().Point().Base())
+	if !left.Equal(right) {
+		return errors.New("bls: invalid signature")
+	}
+	return nil
+}
+
+// Hash a message to a point on curve G1. XXX: This should be replaced
+// eventually by a proper hash-to-point mapping like Elligator.
+func hash(suite pairing.Suite, msg []byte) kyber.Point {
+	h := suite.Hash()
+	h.Write(msg)
+	digest := h.Sum(nil)
+	x := suite.G1().Scalar().SetBytes(digest)
+	return suite.G1().Point().Mul(x, nil)
```

Comment:
Question: Doesn't this invalidate the security proof? I always thought that the point we're chosing should not have a known `Scalar`, else the security proof fails.

Reply:
It's absolutely not ideal we all agree I think but that's the easiest way to get it done. The thing is it does not return the scalar in question, it creates it and then it gets GC'd so in theory there should not be any problem (if the "Mul" operation is constant time ?).
The best solution would be to implement Elligator mappings for any elliptic curve so we can map random bytes to a point easily and safely... that's longterm...

Reply:
Yes it is not ideal but as long as the implementation does not leak the secret scalar everything should be fine. To fix this we should either implement the `Embed` functionality that is currently missing in our `bn256` code or even better the Elligator map as @nikkolasg mentioned.